### PR TITLE
[rust] upgrade to 1.51

### PR DESCRIPTION
**Pull request by sunshowers**
Upgrading Rust to 1.51 brings about the new resolver, along with a
number of lint changes.

Will switch over to the new resolver, and cleanup our current v2
resolver code, in a future PR.

### CODE REVIEW

**Comment by libra-action** [comment]
**This PR may have modified JSON-RPC server code.**

Breaking changes policy:

1. Chaning JSON-RPC method signatures or removing/modifying fields
   in the response in /V1 will not be accepted.
2. Adding fields is ok.
3. Adding a new JSON-RPC method is ok.

Please ensure you have also done the following to update the docs:

1. Update json-rpc/API-CHANGELOG.md.
2. Add/update type/method documentation under json-rpc/docs.

File: storage/jellyfish-merkle/src/tree_cache/mod.rs [hunk 1]
```diff
@@ -268,17 +265,17 @@ where
     }
 }
 
-impl<'a, R, V> Into<(Vec<HashValue>, TreeUpdateBatch<V>)> for TreeCache<'a, R, V>
+impl<'a, R, V> From<TreeCache<'a, R, V>> for (Vec<HashValue>, TreeUpdateBatch<V>)
```
Comment (metajack) [inline storage/jellyfish-merkle/src/tree_cache/mod.rs:268]: What lint was this from?

File: crypto/crypto/src/hkdf.rs [hunk 2]
```diff
@@ -152,6 +152,7 @@ where
 /// c) small PRK value in HKDF-Expand according to RFC 5869.
 /// d) any other underlying HMAC error.
 #[derive(Clone, Debug, PartialEq, Eq, Error)]
+#[allow(clippy::upper_case_acronyms)]
```
Comment (metajack) [inline crypto/crypto/src/hkdf.rs:155]: Is this needed since you whitelisted this in x.toml?

Comment (sunshowers) [reply crypto/crypto/src/hkdf.rs:155]: It's not, will remove.

File: storage/jellyfish-merkle/src/tree_cache/mod.rs (hunk 1 above)
Comment (sunshowers) [reply storage/jellyfish-merkle/src/tree_cache/mod.rs:268]: This one, new in 1.51: https://rust-lang.github.io/rust-clippy/master/#from_over_into

File: crypto/crypto/src/hkdf.rs (hunk 2 above)
Comment (sunshowers) [reply crypto/crypto/src/hkdf.rs]: Done

**Review by metajack — APPROVED** [review]

**Pull request merged**